### PR TITLE
Translationsfield revision

### PR DESCRIPTION
**Pull request by Valc**
@infograf768 
Here is the PR than revise how was wotking before this field and how i think needs work to optimize resources or keep it within a "core" behaivor.

**Before:**

1. It was adding files not present at core language folders, so, a TT sure never check one of that ones, but seem really unrequired show it, due at the end, a package will never be created.
2. Due now we are not using the 'xx-XX' prefix, the call to scan create multiple duplicated of same language  file when it was combined with the bucle for all tags .  I was adding protection against duplicated from a previous PR but now also is skiped the unerquired scans there.
3. Due no reference language, all tags was called to incude options and, as reference, we only need the 'en-GB' tag, and maybe the target language to catch extra files in the selected translation where we wanna create a package. Does not have many sense for the es-ES display extra files, for example, comming from "fr-FR", or chinese language: I can not create my translation with multiple languages if the es-ES not present, and also the package never will be created if the case is present.

**Now**

1. I have added the refence tag, the target language tag, and the mandatory core files we need include.
2. No more search out of the core language folders
3. No more tags than the required ones.
4. I only i have added the target language to be able to see extra files in that translation package or similar cases.
5. For each option we can select from the list now is added a "class" related with what is it if is compared vs the mandatory files.
6. I have not added the CSS due my bad taste with style matters, i am afraid to try it :)

Here the result:
TranslationsField_now

### CODE REVIEW

**Comment by Valc** [comment]
https://github.com/Valc/j4localise/commit/d95ffd1d0771bbfdcc7d0ad9827ba82e9cc8eb2d
I go to sleep, firts time i send by mistake a beta 2 TranslationsField file

**Comment by Valc** [comment]
Last mod adding "Site"

**Comment by infograf768** [comment]
Hmm, don't know what happened, When I want to create a new Extension package, no files are proposed.
I use last master. With and without this patch which btw works fine for core packs.

**Comment by infograf768** [comment]
I found the issue. It comes from my taking off the $tag from the non-core ini files (com_localise and sampledata/testing/).

This line:
https://github.com/infograf768/j4localise/blob/318d6d9ee4231787eb3ba80db1056a89a3af5543/administrator/components/com_localise/Field/ExtensiontranslationsField.php#L126

Taking off `$tag` to get
`$file = "$path$extension/language/$tag/$prefix$extension$suffix.ini";`

solves the issue as we now anyway modify all ini files to take off the $tag.
What do you think?

**Comment by Valc** [comment]
@infograf768 
I think is always without tag, Note that after your PR reformating non core files we will always found it without tag.

**Comment by Valc** [comment]
@infograf768 

Please, test this code. It is almost same revision but within a non-core behaivor

```
<?php
/**
 * @package     Com_Localise
 * @subpackage  models
 *
 * @copyright   Copyright (C) 2005 - 2015 Open Source Matters, Inc. All rights reserved.
 * @license     GNU General Public License version 2 or later; see LICENSE.txt
 */
namespace Joomla\Component\Localise\Administrator\Field;

defined('_JEXEC') or die;

use Joomla\CMS\Component\ComponentHelper;
use Joomla\CMS\Filesystem\File;
use Joomla\CMS\Filesystem\Folder;
use Joomla\CMS\Form\Field\GroupedlistField;
use Joomla\CMS\HTML\HTMLHelper;
use Joomla\CMS\Language\Text;
use Joomla\Component\Localise\Administrator\Helper\LocaliseHelper;
use Joomla\Utilities\ArrayHelper;

/**
 * Form Field ExtensionTranslations class.
 *
 * @package     Extensions.Components
 * @subpackage  Localise
 *
 * @since       1.0
 */
class ExtensiontranslationsField extends GroupedlistField
{
	/**
	 * The field type.
	 *
	 * @var    string
	 */
	protected $type = 'Extensiontranslations';

	/**
	 * Method to get a list of options for a list input.
	 *
	 * @return  array    An array of JHtml options.
	 */
	protected function getGroups()
	{
		$params         = ComponentHelper::getParams('com_localise');
		$reftag	        = !empty($params->get('reference') ? $params->get('reference') : 'en-GB');

		$formdata       = $this->form->getData();
		$langtag        = $formdata["language"];

		$istranslation  = $reftag != $langtag;

		$coreadminfiles = array();
		$coresitefiles  = array();
		$noncorefiles   = array();
		$extrafiles     = array();

		$requiredtags   = array($reftag);

		if ($istranslation)
		{
			$requiredtags[] = $langtag;
		}

		// Remove '.ini' from values
		if (is_array($this->value))
		{
			foreach ($this->value as $key => $val)
			{
				$ext = File::getExt($val);

				if ($ext == "ini")
				{
					$this->value[$key] = substr($val, 0, -4);
				}
			}
		}

		$xml            = simplexml_load_file(JPATH_ROOT . '/media/com_localise/packages/core.xml');
		$coreadminfiles = (array) $xml->administrator->children();
		$coresitefiles  = (array) $xml->site->children();

		$coresitefiles  = $coresitefiles['filename'];
		$coreadminfiles = $coreadminfiles['filename'];

		$coreadminfiles = self::suffix_array_values($coreadminfiles, '.ini');
		$coresitefiles  = self::suffix_array_values($coresitefiles, '.ini');

		$package = (string) $this->element['package'];
		$groups  = array('Site' => array(), 'Administrator' => array());

		$scans = LocaliseHelper::getScans();

		foreach ($scans as $scan)
		{
			$prefix     = $scan['prefix'];
			$suffix     = $scan['suffix'];
			$type       = $scan['type'];
			$client     = ucfirst($scan['client']);
			$path       = $scan['path'];
			$folder     = $scan['folder'];
			$extensions = Folder::folders($path);

			foreach ($extensions as $extension)
			{
				// Take off core extensions
				$file = "$prefix$extension$suffix.ini";
				if (($client == 'Site' && !in_array($file, $coresitefiles)) || ($client == 'Administrator' && !in_array($file, $coreadminfiles)))
				{
					if (Folder::exists("$path$extension/language"))
					{
						// Scan extensions folder
						$tags = Folder::folders("$path$extension/language");

						foreach ($tags as $tag)
						{
							if (!in_array($tag, $requiredtags))
							{
								continue;
							}

							$file   = "$path$extension/language/$tag/$prefix$extension$suffix.ini";

							//Getting the $origin to avoid add, for example, overrides
							$origin = LocaliseHelper::getOrigin("$prefix$extension$suffix", strtolower($client));

							if (File::exists($file) && $origin == '_thirdparty')
							{
								if (!array_key_exists("$prefix$extension$suffix", $groups[$client]))
								{
									$groups[$client]["$prefix$extension$suffix"] = HTMLHelper::_(
											'select.option', strtolower($client) . '_' . "$prefix$extension$suffix", "$prefix$extension$suffix", 'value', 'text', false
									);

									$groups[$client]["$prefix$extension$suffix"]->class = 'non-core-type-' . $type;
								}
							}
						}
					}
				}
			}
		}

		foreach ($groups as $client => $extensions)
		{
			if (count($groups[$client]) == 0)
			{
				$groups[$client][] = HTMLHelper::_('select.option', '',  Text::_('COM_LOCALISE_NOTRANSLATION'), 'value', 'text', true);
			}
			else
			{
				ArrayHelper::sortObjects($groups[$client], 'text');
			}
		}

		// Merge any additional options in the XML definition.
		$groups = array_merge(parent::getGroups(), $groups);

		return $groups;
	}

	/**
	 * Method to add a suffix to an array.
	 *
	 * @param   array   $array   An array of core files.
	 * @param   string  $suffix  The suffix to add to each file.
	 *
	 * @return  array   The modified array
	 */
	public static function suffix_array_values($array, $suffix = '')
	{
		if (!is_array($array))
		{
			return false;
		}

		// Suffix the values and respect the keys
		foreach ($array as $key => $value)
		{
			if (!is_string($value))
			{
				continue;
			}

			$array[$key] = $value . $suffix;
		}

		return $array;
	}
}
```

**Comment by infograf768** [comment]
@valc Please test https://github.com/infograf768/j4localise/pull/27

**Comment by Valc** [comment]
@infograf768
Now is added a litle bit of ajax call when selected language is changed.
Also added last "visual" css part such as wuth non core packs

File: administrator/components/com_localise/Controller/PackageController.php [hunk 1]
```diff
@@ -134,4 +136,47 @@ public function uploadOtherFile()
...
+			$reply->html = $html;
+
+			//If required send a notice apply it before echo
+     		$app->enqueueMessage(Text::_('COM_LOCALISE_UPDATE_TRANSLATIONS_LIST_TASK_SUCCESS'), 'notice');
+
+			echo new JsonResponse($reply, 'Done!');
+		}
+		catch(Exception $e)
+		{
+			$app->enqueueMessage(JText::sprintf('COM_LOCALISE_UPDATE_TRANSLATIONS_LIST_TASK_ERROR', $e->getMessage()), 'error');
```
Comment (infograf768) [inline administrator/components/com_localise/Controller/PackageController.php:172]: JText => Text

File: administrator/components/com_localise/Controller/PackageController.php [hunk 2]
```diff
@@ -134,4 +136,47 @@ public function uploadOtherFile()
...
 		$this->setRedirect(Route::_($url, false));
 	}
+	/**
+	 * Method to update the translarions field list by AJAX call.
+	 *
+	 * @return object
+	 */
+    public function updatetranslationslist()
+    {
+		//Case invalid token, ajax call die here.
```
Comment (infograf768) [inline administrator/components/com_localise/Controller/PackageController.php:146]: space after //

File: administrator/components/com_localise/Controller/PackageController.php [hunk 3]
```diff
@@ -134,4 +136,47 @@ public function uploadOtherFile()
...
+		try
+		{
+			$data  = $input->get('data', null, 'RAW');
+			$data  = json_decode($data);
+
+			$reply->success_message = Text::_('COM_LOCALISE_UPDATE_TRANSLATIONS_LIST_TASK_FLASH_SUCCESS');
+
+			$model = $this->getModel();
+			$html = $model->updateTranslationsList($data);
+			$reply->html = $html;
```
Comment (infograf768) [inline administrator/components/com_localise/Controller/PackageController.php:163]: align = for the 3 lines

File: administrator/components/com_localise/Controller/PackageController.php [hunk 4]
```diff
@@ -134,4 +136,47 @@ public function uploadOtherFile()
...
+
+			//If required send a notice apply it before echo
```
Comment (infograf768) [inline administrator/components/com_localise/Controller/PackageController.php:165]: Space after //

File: administrator/components/com_localise/Controller/PackageController.php [hunk 5]
```diff
@@ -13,6 +13,8 @@
 
 use Joomla\CMS\Factory;
 use Joomla\CMS\Language\Text;
+use Joomla\CMS\Response\JsonResponse;
+use Joomla\CMS\Session\Session;
 use Joomla\CMS\MVC\Controller\FormController;
 use Joomla\CMS\Router\Route;
```
Comment (infograf768) [inline administrator/components/com_localise/Controller/PackageController.php:19]: Please alpha order the `use`

File: administrator/components/com_localise/Controller/PackageController.php (hunk 4 above)
Comment (infograf768) [reply administrator/components/com_localise/Controller/PackageController.php:165]: Comment confusing

File: administrator/components/com_localise/Field/TranslationsField.php [hunk 6]
```diff
@@ -157,4 +244,33 @@ protected function getGroups()
...
+	 * @return  array   The modified array
+	 */
+	public static function suffix_array_values($array, $suffix = '')
+	{
+		if (!is_array($array))
+		{
+			return false;
+		}
+
+		// Suffix the values and respect the keys
```
Comment (infograf768) [inline administrator/components/com_localise/Field/TranslationsField.php:263]: `Suffix the values` means what?

**Comment by Valc** [comment]
I will apply ASAP :)

File: administrator/components/com_localise/Model/PackageModel.php [hunk 7]
```diff
@@ -1331,4 +1333,101 @@ public function uploadOtherFile($file, $location)
...
+	{
+		$params = ComponentHelper::getParams('com_localise');
+		$reftag	= $params->get('reference', '');
+
+		if (empty($reftag))
+		{
+			$reftag = 'en-GB';
+		}
+
+		//Getting the data from ajax call
```
Comment (infograf768) [inline administrator/components/com_localise/Model/PackageModel.php:1352]: Space after //

File: administrator/components/com_localise/Model/PackageModel.php [hunk 8]
```diff
@@ -1331,4 +1333,101 @@ public function uploadOtherFile($file, $location)
...
+		$packagename = htmlspecialchars($data[0]->packagename);
+		$langtag     = htmlspecialchars($data[0]->languagetag);
+
+        //Initiating form instance
+        $filepath = JPATH_ADMINISTRATOR . "/components/com_localise/forms/package.xml";
+        $form_package = Form::getInstance("package", $filepath, array("control" => "jform"));
```
Comment (infograf768) [inline administrator/components/com_localise/Model/PackageModel.php:1358]: space too much after tab for the 3 lines above + missing space after //

File: administrator/components/com_localise/Model/PackageModel.php [hunk 9]
```diff
@@ -1331,4 +1333,101 @@ public function uploadOtherFile($file, $location)
...
+
+        //Form::addFieldPath(JPATH_ADMINISTRATOR . '/components/com_localise/Field');
+        //Form::addFormPath(JPATH_ADMINISTRATOR . '/components/com_localise/forms');
+
+		//Adding next params at translations field only when Ajax call.
```
Comment (infograf768) [inline administrator/components/com_localise/Model/PackageModel.php:1363]: space missing after //

File: administrator/components/com_localise/Model/PackageModel.php [hunk 10]
```diff
@@ -1331,4 +1333,101 @@ public function uploadOtherFile($file, $location)
...
+		$form_package->setFieldAttribute($name = 'translations', 'reftag', $reftag);
+		$form_package->setFieldAttribute($name = 'translations', 'langtag', $langtag);
+
+		$html_output = $form_package->renderField('translations');
+
+		//Cases to "selected" from the xml file if the package filename exists
```
Comment (infograf768) [inline administrator/components/com_localise/Model/PackageModel.php:1369]: Space missing after //
Confusing comment

File: administrator/components/com_localise/Model/PackageModel.php [hunk 11]
```diff
@@ -1331,4 +1333,101 @@ public function uploadOtherFile($file, $location)
...
+							{
+								$content = 'value="administrator_' . $key .'"';
+								$newcontent = 'value="administrator_' . $key .'" selected="selected" ';
+								$lines[$index] = str_replace($content, $newcontent, $line);
+							}
+						}
+					}
+				}
+
+
```
Comment (infograf768) [inline administrator/components/com_localise/Model/PackageModel.php:1403]: take off these 2 empty lines

File: administrator/components/com_localise/Field/TranslationsField.php (hunk 6 above)
Comment (Valc) [reply administrator/components/com_localise/Field/TranslationsField.php:263]: > `Suffix the values` means what?

// Suffix the values and respect the keys

That comment was already rhere.
Means transform com_localise to com_localise.ini or com_localise.sys to com_localise.sys.ini if the suffix from the function is "ini"
All that fine names are stored within an array, so, the function apply that to all matches within the array.

Job time, i will revise all when come back at home.

File: administrator/components/com_localise/language/en-GB/com_localise.ini [hunk 12]
```diff
@@ -158,7 +164,6 @@ COM_LOCALISE_PURGE="Purge Database"
 COM_LOCALISE_PURGE_FAILED="Failed to purge the localise table."
 COM_LOCALISE_PURGE_SUCCESS="The localise table has been emptied successfully."
 COM_LOCALISE_REFORMAT="Reformat j3 languages"
-COM_LOCALISE_REFORMAT_EXTENSION_SUCCESS="The \"%1$s\" ini files have been reformatted." ; The variable is the path to the extension
```
Comment (infograf768) [inline administrator/components/com_localise/language/en-GB/com_localise.ini:161]: Why delete this?

File: administrator/components/com_localise/language/en-GB/com_localise.ini [hunk 13]
```diff
@@ -399,11 +404,12 @@ COM_LOCALISE_ERROR_PACKAGE_XML="The xml file is not correctly formatted"
 COM_LOCALISE_ERROR_PACKAGES_EXPORT="Error exporting the package xml. The file may not exist or you try to export reference package."
 COM_LOCALISE_EXPORT_NOT_PERMITTED="You are not authorised to export a package xml'"
 COM_LOCALISE_FIELDSET_COREPACKAGE_TRANSLATIONS="Choose only the core files translations"
+COM_LOCALISE_FIELDSET_COREPACKAGE_TRANSLATIONS_DESC="<span class='missing'>Without translation for the language selected.</span> <span class='extra'>Extra in translation for the language selected.</span>"
 COM_LOCALISE_FIELDSET_PACKAGE_DETAIL="Core Package Details"
 COM_LOCALISE_FIELDSET_PACKAGE_DETAIL_DESC="Core Package details description"
 COM_LOCALISE_FIELDSET_PACKAGE_PERMISSIONS="Permissions"
 COM_LOCALISE_FIELDSET_PACKAGE_TRANSLATIONS="Translations"
-COM_LOCALISE_FIELDSET_PACKAGE_TRANSLATIONS_DESC="Translations description"
+COM_LOCALISE_FIELDSET_PACKAGE_TRANSLATIONS_DESC="<span class='missing'>Without translation for the language selected.</span> <span class='extra'>Extra in translation for the language selected.</span>"
```
Comment (infograf768) [inline administrator/components/com_localise/language/en-GB/com_localise.ini:412]: Would have to test but what about:
<span class='missing'>Missing translation file in the selected language.</span> <span class='extra'>Extra translation file in the selected language.</span>"

File: administrator/components/com_localise/Field/TranslationsField.php [hunk 14]
```diff
@@ -44,6 +45,49 @@ class TranslationsField extends GroupedlistField
...
+		$reftag	        = $params->get('reference', '');
+
+		if (empty($reftag))
+		{
+			$reftag = 'en-GB';
+		}
+
+		//Form priority
+		$formdata       = $this->form->getData();
+		$langtag        = $formdata["language"];
```
Comment (infograf768) [inline administrator/components/com_localise/Field/TranslationsField.php:60]: lines 60, 59, 51, 50 too much spaces before the =

File: administrator/components/com_localise/Field/TranslationsField.php [hunk 15]
```diff
@@ -44,6 +45,49 @@ class TranslationsField extends GroupedlistField
...
+
+		//Ajax priority
```
Comment (infograf768) [inline administrator/components/com_localise/Field/TranslationsField.php:62]: space after //

**Comment by infograf768** [comment]
As our CSS is very small, I see no reason to create a new package.css

**Comment by infograf768** [comment]
BTW, just corrected a missing class. See https://github.com/infograf768/j4localise/commit/95c24882d7b5466e123b7250421ed8e3f539a96d

File: administrator/components/com_localise/Field/TranslationsField.php (hunk 6 above)
Comment (infograf768) [reply administrator/components/com_localise/Field/TranslationsField.php:263]: I suggest therefore
`Add a suffix to the value...`

File: administrator/components/com_localise/language/en-GB/com_localise.ini [hunk 16]
```diff
@@ -36,6 +36,12 @@ ADMINISTRATOR="Administrator"
 SITE="Site"
 INSTALLATION="Installation"
 
+[Ajax]
+
+COM_LOCALISE_UPDATE_TRANSLATIONS_LIST_TASK_FLASH_ERROR="Error updating Translations list"
+COM_LOCALISE_UPDATE_TRANSLATIONS_LIST_TASK_FLASH_SUCCESS="Translations list has been updated"
+COM_LOCALISE_UPDATE_TRANSLATIONS_LIST_TASK_SUCCESS="Translations list has been updated."
```
Comment (infograf768) [inline administrator/components/com_localise/language/en-GB/com_localise.ini:43]: Do we really need 2 strings with the same value?

File: administrator/components/com_localise/language/en-GB/com_localise.ini (hunk 13 above)
Comment (infograf768) [reply administrator/components/com_localise/language/en-GB/com_localise.ini:412]: Also, I would prefer  to separate Missing and Extra with a <br>

**Comment by infograf768** [comment]
I would prefer using `<p>` instead of `<legend>` a second time in order to get rid of the bold. Also separate both span with `<br>`

`COM_LOCALISE_FIELDSET_COREPACKAGE_TRANSLATIONS_DESC="<span class='missing'>Missing translation file in the selected language.</span><br><span class='extra'>Extra translation file in the selected language.</span>"`

```
								<legend><?php echo Text::_($fieldSets['translations']->label); ?></legend>
								<?php if (!empty($fieldSets['translations']->description)):?>
										<p><?php echo Text::_($fieldSets['translations']->description); ?></p>
								<?php endif;?>
```

Screen Shot 2021-09-17 at 10 17 48

File: administrator/components/com_localise/language/en-GB/com_localise.ini (hunk 12 above)
Comment (Valc) [reply administrator/components/com_localise/language/en-GB/com_localise.ini:161]: Utch, by mistake.
Seems was using an not updated language file

File: administrator/components/com_localise/language/en-GB/com_localise.ini (hunk 13 above)
Comment (Valc) [reply administrator/components/com_localise/language/en-GB/com_localise.ini:412]: Done

File: administrator/components/com_localise/language/en-GB/com_localise.ini (hunk 16 above)
Comment (Valc) [reply administrator/components/com_localise/language/en-GB/com_localise.ini:43]: I think yes due the "flash" ones should be a short text easy to read fast due there is only 3 seconds to read it.
The not "flash" can be longuer text due are the "system messages", displayed at the top of the page.

The text now is similar due no time to improve it yesterday, now will come updated at next PR

**Comment by Valc** [comment]
@infograf768 
Nice output now :)
> Screen Shot 2021-09-17 at 10 17 48

I think i have aookied all your sugestions now.

Also added at model a new case:
```
// Useful when creating a new core package that does not have a package name assigned
// In that case, the translations to be highlighted from the list will be the set by mandatory from core
// when selected language is changed and run this ajax call.
```
We are handling close to 400 files

To test:
Open from "Packages" view a "New core package"

No "package name" and not highlighted fields there now.

Then, select "en-GB" or any other not showed as default.

Expected result: the mandatory "en-GB" files set is automatically highligted.

Then you can save that one as "xx-XX language pack version" matching that core.xml set

File: administrator/components/com_localise/Controller/PackageController.php [hunk 17]
```diff
@@ -134,4 +136,59 @@ public function uploadOtherFile()
 
 		$this->setRedirect(Route::_($url, false));
 	}
+	/**
+	 * Method to update the translarions field list by AJAX call.
```
Comment (infograf768) [inline administrator/components/com_localise/Controller/PackageController.php:140]: translations, not translarions

File: administrator/components/com_localise/Field/TranslationsField.php [hunk 18]
```diff
@@ -157,4 +244,36 @@ protected function getGroups()
...
+	{
+		if (!is_array($array))
+		{
+			return false;
+		}
+
+		// Add a suffix to the values without change the keys.
+		// For example when $suffix = '.ini', if the stored key in array is "com_localise" the asigned value to that key will change to "com_localise.ini".
+		// Usefull when we get the "$xml = simplexml_load_file(JPATH_ROOT . '/media/com_localise/packages/core.xml');"
+		// due at "core.xml" file, the file names have not the '.ini' suffix added.
```
Comment (infograf768) [inline administrator/components/com_localise/Field/TranslationsField.php:266]: suggest

```
		// Add a suffix to the values without changing the keys.
		// For example when $suffix = '.ini', if the stored key in array is "com_localise", the assigned value for that key will be changed to "com_localise.ini".
		// Useful when we get the "$xml = simplexml_load_file(JPATH_ROOT . '/media/com_localise/packages/core.xml');"
		// as the files names in "core.xml" have no '.ini' suffix.
		```

File: administrator/components/com_localise/Model/PackageModel.php [hunk 19]
```diff
@@ -1331,4 +1333,157 @@ public function uploadOtherFile($file, $location)
...
+	/**
+	 * Method to get the HTML output required to update the translations field list by the selected language tag.
+	 *
+	 * @return   string  The data for the jform_translations field.
+	 */
+	public function updateTranslationsList($data)
+	{
+		$params = ComponentHelper::getParams('com_localise');
+		$reftag	= $params->get('reference', '');
```
Comment (infograf768) [inline administrator/components/com_localise/Model/PackageModel.php:1345]: take off one space before =

File: administrator/components/com_localise/Model/PackageModel.php [hunk 20]
```diff
@@ -1331,4 +1333,157 @@ public function uploadOtherFile($file, $location)
...
+		$langtag     = htmlspecialchars($data[0]->languagetag);
+
+		// Initiating form instance
+		$filepath = JPATH_ADMINISTRATOR . "/components/com_localise/forms/package.xml";
+		$form_package = Form::getInstance("package", $filepath, array("control" => "jform"));
+
+		//Form::addFieldPath(JPATH_ADMINISTRATOR . '/components/com_localise/Field');
+		//Form::addFormPath(JPATH_ADMINISTRATOR . '/components/com_localise/forms');
+
+		// Adding the params bellow at translations field only when Ajax call.
```
Comment (infograf768) [inline administrator/components/com_localise/Model/PackageModel.php:1363]: bellow => below

`// Adding the params below for the translations field only when Ajax call.`

File: administrator/components/com_localise/Model/PackageModel.php [hunk 21]
```diff
@@ -1331,4 +1333,157 @@ public function uploadOtherFile($file, $location)
...
+
+			$xml   = simplexml_load_file(JPATH_ROOT . '/media/com_localise/packages/core.xml');
+			$lines = preg_split("/\\r\\n|\\r|\\n/", $html_output);
+
+			if ($xml->administrator)
+			{
+				foreach ($xml->administrator->children() as $file)
+				{
+					// core.xml file never have the '.ini' added at file names, so, in this case, is not required delete the '.ini'.
+					//$key    = substr($file, 0, strlen($file) - 4);
```
Comment (infograf768) [inline administrator/components/com_localise/Model/PackageModel.php:1442]: I guess no use to keep these 2 comments

File: administrator/components/com_localise/language/en-GB/com_localise.ini [hunk 22]
```diff
@@ -36,6 +36,13 @@ ADMINISTRATOR="Administrator"
 SITE="Site"
 INSTALLATION="Installation"
 
+[Ajax]
+
+COM_LOCALISE_UPDATE_TRANSLATIONS_LIST_TASK_ERROR="Unable to update the Translations list."
+COM_LOCALISE_UPDATE_TRANSLATIONS_LIST_TASK_FLASH_ERROR="Error updating Translations list"
+COM_LOCALISE_UPDATE_TRANSLATIONS_LIST_TASK_FLASH_SUCCESS="Translations list has been updated"
+COM_LOCALISE_UPDATE_TRANSLATIONS_LIST_TASK_SUCCESS="The file list for the selected language has been updated. Before 'Save' it would be advisable to check that the name of the package and the name of the selected language correspond."
```
Comment (infograf768) [inline administrator/components/com_localise/language/en-GB/com_localise.ini:44]: suggest

`"The file list for the selected language has been updated. Make sure that the package name and the selected language correspond before saving."`

File: administrator/components/com_localise/Model/PackageModel.php (hunk 19 above)
Comment (Valc) [reply administrator/components/com_localise/Model/PackageModel.php:1345]: ??? i can not to see that one. If posible, please, paste here the right one.

File: administrator/components/com_localise/Model/PackageModel.php (hunk 20 above)
Comment (Valc) [reply administrator/components/com_localise/Model/PackageModel.php:1363]: Done

File: administrator/components/com_localise/Model/PackageModel.php (hunk 21 above)
Comment (Valc) [reply administrator/components/com_localise/Model/PackageModel.php:1442]: Done

File: administrator/components/com_localise/language/en-GB/com_localise.ini (hunk 22 above)
Comment (Valc) [reply administrator/components/com_localise/language/en-GB/com_localise.ini:44]: Done

File: administrator/components/com_localise/Field/TranslationsField.php (hunk 18 above)
Comment (Valc) [reply administrator/components/com_localise/Field/TranslationsField.php:266]: Done

File: administrator/components/com_localise/Model/PackageModel.php (hunk 19 above)
Comment (Valc) [reply administrator/components/com_localise/Model/PackageModel.php:1345]: Done

File: administrator/components/com_localise/Controller/PackageController.php (hunk 17 above)
Comment (Valc) [reply administrator/components/com_localise/Controller/PackageController.php:140]: Done

File: administrator/components/com_localise/Model/PackageModel.php (hunk 19 above)
Comment (infograf768) [reply administrator/components/com_localise/Model/PackageModel.php:1345]: strange. It shows here above but not in a text file... Forget it. :)

**Comment by Valc** [comment]
@infograf768 
Suggestions applied

**Comment by infograf768** [comment]
1. Suggestion:
Change xml language field to
```
		<field
			id="language"
			name="language"
			type="corelanguage"
			label="COM_LOCALISE_LABEL_PACKAGE_LANGUAGE"
			required="true">
			<option value="">COM_LOCALISE_OPTION_LANGUAGE_SELECT</option>
		</field>
```
Otherwise the first alpha ordered language will display and no list highlighted as it needs to first use Ajax to get the list.
(Will also change to a Select for packagefile to force selecting the correct lang)

2. Problem at attempting first Save

Screen Shot 2021-09-18 at 10 10 40
A second Save is fine.

Once this is solved, I wonder if we should keep the Master Core package.

**Comment by Valc** [comment]
@infograf768 
> 1. Change xml language field to

Yeah, it solve easy and fast!!
Applied!

About that message, i had alrredy noticed, but not sure about how to solve.
https://github.com/infograf768/j4localise/pull/17
Then i say:
`Finally pray that someone knows how to help with the "path" issse: i have no idea.`

Now i think this one can to be directly related whe the table is created by "installation.xml" file due the "path" have not asigned a default vaue.
But at the end, i does not know if is required ad as "deafault => null" or other right to the case  or if really is comming all from there.

Maybe you or @bakual can hit better than me the case.

**Comment by Bakual** [comment]
I'm not exactly sure what database table you're talking about here, but if the field has "NOT NULL" you indeed need to provide a value for it when saving the row. Or you get that "Field path doesn't have a default value" error.
Best way is to add that "path" property to the table object and assign a value before saving it. A default value in the database table itself would be more a bandaid since the default value usually will not be the correct value.

**Comment by infograf768** [comment]
We speak about the localise table.

Screen Shot 2021-09-18 at 18 33 25

Example:

Screen Shot 2021-09-18 at 18 35 24

**Comment by Valc** [comment]
Such as @Bakual explain i understand the best option is set the "path" from the code side, and not changing the table structure created from our install.sql file.

Seems than if the crash begin just after "Save" function, it will be solved from "save" function at "PackageModel" file.

After some "die" i have noticed than when a new package is created the "$olpath" is always "NULL" and this case seems is not handled from there.

Now i go to test as first attemp this one close to the end of the function:

```
		if ($path == $oldpath)
		{
			$id = LocaliseHelper::getFileId($path);
			$this->setState('package.id', $id);

			// Bind the rules.
			$table = $this->getTable();
			$table->load($id);
		}
		elseif (!empty($path) && $oldpath == NULL)
		{
			$table = $this->getTable();

			$table->store();

			$id = LocaliseHelper::getFileId($path);
			$this->setState('package.id', $id);
			$app->setUserState('com_localise.edit.package.id', $id);
		}
		else
		{
			$table = $this->getTable();

			if (!$table->delete((int) $originalId))
			{
				$this->setError($table->getError());

				return false;
			}

			$table->store();

			$id = LocaliseHelper::getFileId($path);
			$this->setState('package.id', $id);
			$app->setUserState('com_localise.edit.package.id', $id);
		}
```

The new one is:

```
		elseif (!empty($path) && $oldpath == NULL)
		{
			$table = $this->getTable();

			$table->store();

			$id = LocaliseHelper::getFileId($path);
			$this->setState('package.id', $id);
			$app->setUserState('com_localise.edit.package.id', $id);
		}
```

Still not tested i have 5 minutes time to be waitting before crash aaaaall!

But if not works, i think is close to here how to solve

**Comment by Valc** [comment]
@infograf768 
Edited due i think is solved now.
Please, JM. Test it.

@Bakual Thank you! but if you notice do that solution is crazy.. please, bring more light :)

**Comment by infograf768** [comment]
the part from `or with` works fine here.
If no more feedback/suggestion from @Bakual please modify PR and let's remember that we have to do the same for packagefile.

**Comment by Valc** [comment]
@infograf768
Done

**Comment by Valc** [comment]
@infograf768 
Please, wait a bit before send the PR.
I am improving a bit more the Ajax query function from edit view.
Now its not noticed, for, example, when session is expired.
I will have ready today.
That new structure at the function will allow handle all type of cases.
So, it will be usefull here and at packagefile controler but also at other PR i have in mind send to you, such as add filters or be able to save "key by key" features.

**Comment by Valc** [comment]
@infograf768 
Ready.

session

false

**Comment by Valc** [comment]
Now working on "non-core" implementation. Its a bit more complex but will send when done.

**Comment by infograf768** [comment]
Concerning packagefile, I better close https://github.com/infograf768/j4localise/pull/27 and let you propose a full PR for it, after this one here is merged. Give me some time to test this now

File: media/com_localise/css/localise.css [hunk 23]
```diff
@@ -349,3 +349,34 @@ color:#8B0000 !important;
 .info-line {
   line-height: 1.2rem;
 }
+/* TanslationsField list */
```
Comment (infograf768) [inline media/com_localise/css/localise.css:352]: TranslationsField, not TanslationsField

File: administrator/components/com_localise/language/en-GB/com_localise.ini [hunk 24]
```diff
@@ -36,6 +36,15 @@ ADMINISTRATOR="Administrator"
 SITE="Site"
 INSTALLATION="Installation"
 
+[Ajax]
+
+COM_LOCALISE_TASK_UPDATE_TRANSLATIONS_LIST_ERROR="Unable to update the Translations list."
+COM_LOCALISE_TASK_UPDATE_TRANSLATIONS_LIST_ERROR_FLASH="Error updating Translations list"
```
Comment (infograf768) [inline administrator/components/com_localise/language/en-GB/com_localise.ini:42]: Error updating `the` translations list

also please take off uppercase for `translations` here and the line above

**Comment by Valc** [comment]
> Concerning packagefile, I better close #27 and let you propose a full PR for it, after this one here is merged. Give me some time to test this now

Yes, i think also is better close #27. Please, test it all time you need :)

In a bit i will send the suggestions applied and i think the other PR is also ready to go.

**Comment by infograf768** [comment]
@Valc 

What is `COM_LOCALISE_OPTION_EXTENSION_NAME_SELECT` doing here?

**Comment by Valc** [comment]
@infograf768 
Uf, seems i have mixed that content at language file that was ready to be use at package file revision PR. Att the end, there i have added a "Packagefile" section and i forguet move the COM_LOCALISE_OPTION_EXTENSION_NAME_SELECT key there.

I will update the language file in a bit deleting that extra keys for this PR

**Comment by infograf768** [comment]
Warning
The other PR also contains this one...

**Comment by Valc** [comment]
@infograf768 
> Warning
> The other PR also contains this one...

Arrrrrrrgh! i have mixed both PRs there.
Seems i start the new PR forguering first check out to master branch and then create a new branck to handle the new PR.
I will tray to solve it, but i think i need close the other and estart a new one with only the files to change.

**Comment by infograf768** [comment]
Merged, thanks.
Please apply to your branch for packagefile.

**Pull request merged**